packages/video_player/video_player_android/lib/src/android_video_player.dart

```diff
@@ -280,6 +237,130 @@ PlatformVideoViewType _platformVideoViewTypeFromVideoViewType(
+    return _api.setPlaybackSpeed(speed);
+  }
+
+  Future<void> seekTo(Duration position) {
+    return _api.seekTo(position.inMilliseconds);
+  }
+
+  Future<Duration> getPosition() async {
+    final PlaybackState state = await _api.getPlaybackState();
+    // TODO(stuartmorgan): Move this logic. This is a workaround for the fact
+    //  that ExoPlayer doesn't have any way to observe buffer position
+    //  changes, so polling is required. To minimize platform channel overhead,
+    //  that's combined with getting the position, but this relies on the fact
+    //  that the app-facing package polls getPosition frequently, which makes
+    //  this fragile (for instance, as of writing, this won't be called while
+    //  the video is paused). It should instead be called on its own timer,
+    //  independent of higher-level package logic.
+    _updateBufferingState(state.bufferPosition);
+    return Duration(milliseconds: state.playPosition);
+  }
+
+  Stream<VideoEvent> videoEvents() {
+    return _eventStreamController.stream;
+  }
+
+  Future<void> dispose() async {
+    await _eventSubscription.cancel();
+  }
+
+  /// Sends a buffering update if the buffer position has changed since the
+  /// last check.
+  void _updateBufferingState(int bufferPosition) {
+    if (bufferPosition != _lastBufferPosition) {
+      _lastBufferPosition = bufferPosition;
+      _eventStreamController.add(
+        VideoEvent(
+          eventType: VideoEventType.bufferingUpdate,
+          buffered: _bufferRangeForPosition(bufferPosition),
+        ),
+      );
+    }
+  }
+
+  void _onStreamEvent(dynamic event) {
+    final Map<dynamic, dynamic> map = event as Map<dynamic, dynamic>;
+    _eventStreamController.add(switch (map['event']) {
+      'initialized' => VideoEvent(
+        eventType: VideoEventType.initialized,
+        duration: Duration(milliseconds: map['duration'] as int),
+        size: Size(
+          (map['width'] as num?)?.toDouble() ?? 0.0,
+          (map['height'] as num?)?.toDouble() ?? 0.0,
+        ),
+        rotationCorrection: map['rotationCorrection'] as int? ?? 0,
+      ),
+      'completed' => VideoEvent(eventType: VideoEventType.completed),
+      'bufferingUpdate' => VideoEvent(
+        eventType: VideoEventType.bufferingUpdate,
+        buffered: _bufferRangeForPosition(map['position'] as int),
+      ),
+      'bufferingStart' => VideoEvent(eventType: VideoEventType.bufferingStart),
+      'bufferingEnd' => VideoEvent(eventType: VideoEventType.bufferingEnd),
+      'isPlayingStateUpdate' => VideoEvent(
+        eventType: VideoEventType.isPlayingStateUpdate,
+        isPlaying: map['isPlaying'] as bool,
+      ),
+      _ => VideoEvent(eventType: VideoEventType.unknown),
+    });
+  }
+
+  // Turns a single buffer position, which is what ExoPlayer reports, into the
+  // DurationRange array expected by [VideoEventType.bufferingUpdate].
+  List<DurationRange> _bufferRangeForPosition(int milliseconds) {
```

Comment:
This was extracted from the switch to allow it to be used in `_updateBufferingState` as well.